cmd/src/debug_comp.go

```diff
@@ -0,0 +1,255 @@
+		if err != nil {
+			fmt.Errorf("failed to get containers for subcommand with err: %v", err)
+		}
+		// Safety check user knows what they are targeting with this debug command
+		log.Printf("This command will archive docker-cli data for %d containers\n SRC_ENDPOINT: %v\n Output filename: %v", len(containers), cfg.Endpoint, base)
+		if verified, _ := verify("Do you want to start writing to an archive?"); !verified {
+			return nil
+		}
+
+		err = archiveCompose(ctx, zw, *verbose, configs, baseDir)
+		if err != nil {
+			return cmderrors.ExitCode(1, nil)
+		}
+		return nil
+	}
+
+	debugCommands = append(debugCommands, &command{
+		flagSet: flagSet,
+		handler: handler,
+		usageFunc: func() {
+			fmt.Println(usage)
+		},
+	})
+}
+
+/*
+Docker functions
+*/
+
+func archiveCompose(ctx context.Context, zw *zip.Writer, verbose, configs bool, baseDir string) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	containers, err := getContainers(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get docker containers: %w", err)
+	}
+
+	if verbose {
+		log.Printf("getting docker data for %d containers...\n", len(containers))
+	}
+
+	// setup channel for slice of archive function outputs
+	ch := make(chan *archiveFile)
+	g, ctx := errgroup.WithContext(ctx)
+	semaphore := semaphore.NewWeighted(8)
+
+	// start goroutine to run docker ps -o wide
+	g.Go(func() error {
+		if err := semaphore.Acquire(ctx, 1); err != nil {
+			return err
+		}
+		defer semaphore.Release(1)
+		ch <- getPs(ctx, baseDir)
+		return nil
+	})
+
+	// start goroutine to run docker container stats --no-stream
+	g.Go(func() error {
+		if err := semaphore.Acquire(ctx, 1); err != nil {
+			return err
+		}
+		defer semaphore.Release(1)
+		ch <- getStats(ctx, baseDir)
+		return nil
+	})
+
+	// start goroutine to run docker container logs <container>
+	for _, container := range containers {
+		c := container
+		g.Go(func() error {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return err
+			}
+			defer semaphore.Release(1)
+			ch <- getContainerLog(ctx, c, baseDir)
+			return nil
+		})
+	}
+
+	// start goroutine to run docker container inspect <container>
+	for _, container := range containers {
+		c := container
+		g.Go(func() error {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return err
+			}
+			defer semaphore.Release(1)
+			ch <- getInspect(ctx, c, baseDir)
+			return nil
+		})
+	}
+
+	// start goroutine to get configs
+	if configs {
+		g.Go(func() error {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return err
+			}
+			defer semaphore.Release(1)
+			ch <- getExternalServicesConfig(ctx, baseDir)
+			return nil
+		})
+
+		g.Go(func() error {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return err
+			}
+			defer semaphore.Release(1)
+			ch <- getSiteConfig(ctx, baseDir)
+			return nil
+		})
+	}
+
+	// close channel when wait group goroutines have completed
+	go func() {
+		g.Wait()
```

Comment:
`g.Wait()` returns an error value, but it doesn't seem very significant here. I think the only way for the errgroup to return an error is if the context got canceled, which means the app is exiting anyway.

I think it's safe not to pass this value back, but you should at the very least print it when it occurs just so we're not silently failing in the future.

Passing the value back is a bit more difficult since it's in a goroutine (you'd need to share memory or open a new channel just to receive that value).

Reply:
Added a `fmt.Printf` but going to leave this open till I can discuss